### ISSUE TITLE
Many To Many Polymorphic (referencing the same table) / Alias 

### ISSUE THREAD

**Issue by mikeyb31** [laravel/framework#12219]
laravel/framework v5.1.29

    SQLSTATE[42S22]: Column not found: 1054 Unknown column 'categorables.categorable_type' in 'where clause' (SQL: select count(*) as aggregate from `posts` where `posts`.`post_type` in (page, widget, post, partner, faq, document, category) and `post_type` in (post) and (select count(*) from `categorables` as `self_e8162e2b0a733a7f1a1c229424c68a45` where `self_e8162e2b0a733a7f1a1c229424c68a45`.`categorable_id` = `posts`.`id` and `categorables`.`categorable_type` = Cms\Models\Post and `post_id` in (24) and `posts`.`post_type` in (category)) >= 1 and `status` = 1)
    
Note the alias:
    
    select count(*) from `categorables` as `self_e8162e2b0a733a7f1a1c229424c68a45` where `self_e8162e2b0a733a7f1a1c229424c68a45`.`categorable_id` = `posts`.`id` and `categorables`.`categorable_type` = Cms\Models\Post and `post_id` in (24) and `posts`.`post_type` in (category)
    
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'categorables.categorable_type' in 'where clause'   
Shouldn't the alias `self_4d0b35cab2a3ca21a451d564a8e16761` be prefixed to `categorable_type`?

**Comment by patrickcarlohickman** [comment]
I created PR #12146 which was merged into 5.2 to resolve this type of issue, but I didn't test it with a many to many polymorphic relationship. Can you show your Model and relationship definitions? Thanks.

**Comment by GrahamCampbell** [comment]
Ping @taylorotwell regarding the possibility of having the fix backported to 5.1.

**Comment by patrickcarlohickman** [comment]
I tested the code in PR #12146 with this scenario and I believe it works with this polymorphic many-to-many scenario, as well.

I created a test with the following models and migrations:

**Post.php**

```php
class Post extends Model
{
    public function categories()
    {
        return $this->morphToMany(Category::class, 'categorable');
    }
}
```

**Category.php**

```php
class Category extends Model
{
    public function posts()
    {
        return $this->morphedByMany(Post::class, 'categorable');
    }

    public function subCategories()
    {
        return $this->morphToMany(static::class, 'categorable');
    }

    public function parentCategories()
    {
        return $this->morphedByMany(static::class, 'categorable');
    }
}
```

**Migrations**

```php
// posts
public function up()
{
    Schema::create('posts', function (Blueprint $table) {
        $table->engine = 'InnoDB';

        $table->increments('id');
        $table->timestamps();
        $table->string('title');
        $table->string('type', 50);
    });
}

// categories
public function up()
{
    Schema::create('categories', function (Blueprint $table) {
        $table->engine = 'InnoDB';

        $table->increments('id');
        $table->timestamps();
        $table->string('name', '50');
    });
}

// polymorphic pivot
public function up()
{
    Schema::create('categorables', function (Blueprint $table) {
        $table->engine = 'InnoDB';

        $table->integer('category_id')->unsigned();
        $table->morphs('categorable');
    });
}
```

With the above setup, previous to the PR, the below statement errors out:

```php
\App\Category::has('subCategories')->get();
```

>Illuminate\Database\QueryException with message 'SQLSTATE[42S22]: Column not found: 1054 Unknown column 'categorables.categorable_type' in 'where clause' (SQL: select * from `categories` where exists (select * from `categorables` as `self_8468df4eb330a1f52be1a4559e0673b2` where `self_8468df4eb330a1f52be1a4559e0673b2`.`categorable_id` = `categories`.`id` and `categorables`.`categorable_type` = App\Category))'

Additionally, the below statement does not error but also does not return the correct results:

```php
\App\Post::has('categories.subCategories')->get();
```

After the PR, both above statements work correctly and return the correct results.

**Comment by mikeyb31** [comment]
Hi @patrickcarlohickman 

Thank you so much for replying :)

My models are almost identical to what you posted above and I confirm that when I get a post model and `dd($post->categories)` or the reverse, get a category and `dd($category->posts)` this works absolutely fine confirming that the relationships are setup correctly.

It's only once I try and do a whereHas() that the problem occurs `Column not found`

**Comment by patrickcarlohickman** [comment]
@mikeyb31 As I mentioned, this issue has recently been fixed in 5.2. I have created PR #12312 to have this fix backported into 5.1. You can keep an eye on this new PR to see if/when it is merged. Hope it helps!

**Comment by mikeyb31** [comment]
@patrickcarlohickman Thank you! :)
